File: 2023/day08/zig/solution.zig
const std = @import("std");

const Node = struct {
    left: [3]u8,
    right: [3]u8,
};

fn parseNodeName(slice: []const u8) [3]u8 {
    return .{ slice[0], slice[1], slice[2] };
}

fn lcm(a: u64, b: u64) u64 {
    return a / std.math.gcd(a, b) * b;
}

/// Navigate from a starting node until the predicate returns true.
/// Returns the number of steps taken.
fn navigate(
    network: std.AutoHashMap([3]u8, Node),
    instructions: []const u8,
    start: [3]u8,
    comptime isEnd: fn ([3]u8) bool,
) !u64 {
    var current = start;
    var steps: u64 = 0;

    while (!isEnd(current)) {
        const instruction = instructions[steps % instructions.len];
        const node = network.get(current) orelse return error.NodeNotFound;
        current = if (instruction == 'L') node.left else node.right;
        steps += 1;
    }

    return steps;
}

fn isZZZ(node: [3]u8) bool {
    return std.mem.eql(u8, &node, &[3]u8{ 'Z', 'Z', 'Z' });
}

fn endsWithZ(node: [3]u8) bool {
    return node[2] == 'Z';
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const input = try std.fs.cwd().readFileAlloc(allocator, "../input.txt", 1024 * 1024);
    defer allocator.free(input);

    // Parse input
    var lines = std.mem.splitScalar(u8, input, '\n');
    const instructions = lines.next() orelse return error.InvalidInput;

    // Skip empty line
    _ = lines.next();

    // Parse network
    var network = std.AutoHashMap([3]u8, Node).init(allocator);
    defer network.deinit();

    var starting_nodes: std.ArrayListUnmanaged([3]u8) = .empty;
    defer starting_nodes.deinit(allocator);

    while (lines.next()) |line| {
        // Minimum valid line: "AAA = (BBB, CCC)" is 16 characters
        if (line.len < 16) continue;

        // Parse: AAA = (BBB, CCC)
        const node_name = parseNodeName(line[0..3]);
        const left = parseNodeName(line[7..10]);
        const right = parseNodeName(line[12..15]);

        try network.put(node_name, .{ .left = left, .right = right });

        // Check if it's a starting node (ends with 'A')
        if (node_name[2] == 'A') {
            try starting_nodes.append(allocator, node_name);
        }
    }

    // Part 1: Navigate from AAA to ZZZ
    const part1 = try navigate(network, instructions, .{ 'A', 'A', 'A' }, isZZZ);
    std.debug.print("Part 1: {}\n", .{part1});

    // Part 2: Find LCM of all cycle lengths
    var result: u64 = 1;
    for (starting_nodes.items) |start_node| {
        const cycle_steps = try navigate(network, instructions, start_node, endsWithZ);
        result = lcm(result, cycle_steps);
    }
    std.debug.print("Part 2: {}\n", .{result});
}

File: 2023/day08/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Predicate;

public class Solution {
    private record Node(String left, String right) {}

    private final String instructions;
    private final Map<String, Node> network;

    public Solution(String input) {
        String[] parts = input.split("\n\n");
        this.instructions = parts[0].trim();
        this.network = new HashMap<>();

        for (String line : parts[1].split("\n")) {
            if (line.isBlank()) continue;
            // Parse: AAA = (BBB, CCC)
            String[] nodeParts = line.split(" = ");
            String name = nodeParts[0];
            String connections = nodeParts[1].substring(1, nodeParts[1].length() - 1);
            String[] leftRight = connections.split(", ");
            network.put(name, new Node(leftRight[0], leftRight[1]));
        }
    }

    private long navigate(String start, Predicate<String> isEnd) {
        String current = start;
        long steps = 0;
        int instructionLen = instructions.length();

        while (!isEnd.test(current)) {
            char instruction = instructions.charAt((int) (steps % instructionLen));
            Node node = network.get(current);
            current = (instruction == 'L') ? node.left() : node.right();
            steps++;
        }

        return steps;
    }

    public long part1() {
        return navigate("AAA", node -> node.equals("ZZZ"));
    }

    public long part2() {
        return network.keySet().stream()
                .filter(node -> node.endsWith("A"))
                .mapToLong(start -> navigate(start, node -> node.endsWith("Z")))
                .reduce(1L, Solution::lcm);
    }

    private static long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private static long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }

    public static void main(String[] args) throws IOException {
        String input = Files.readString(Path.of("../input.txt"));
        Solution solution = new Solution(input);

        System.out.println("Part 1: " + solution.part1());
        System.out.println("Part 2: " + solution.part2());
    }
}
